sunbeam-python/sunbeam/provider/local/steps.py

```diff
@@ -255,3 +267,282 @@ def _update_microcluster_status(self, status: dict, microcluster_status: dict):
+                nic["pci_address"],
+            )
+            node_excluded_devices.remove(pci_address)
+
+        # Update the global whitelist if needed.
+        whitelisted = self._is_sriov_nic_whitelisted(
+            nic, pci_whitelist, excluded_devices
+        )[0]
+        if not whitelisted:
+            new_dev_spec = {
+                "address": nic["pci_address"],
+                "vendor_id": nic["vendor_id"].lstrip("0x"),
+                "product_id": nic["product_id"].lstrip("0x"),
+                "physical_network": physnet,
+            }
+            pci_whitelist.append(new_dev_spec)
+        else:
+            LOG.debug(
+                "SR-IOV nic already whitelisted: %s %s", nic["name"], nic["pci_address"]
+            )
+
+    def _exclude_sriov_nic(
+        self, nic: dict, pci_whitelist: list[dict], excluded_devices: dict[str, list]
+    ):
+        LOG.debug("Excluding SR-IOV nic: %s", nic["name"])
+        if self.node_name not in excluded_devices:
+            excluded_devices[self.node_name] = []
+        if nic["pci_address"] not in excluded_devices[self.node_name]:
+            excluded_devices[self.node_name].append(nic["pci_address"])
+
+    def _get_nic_str_repr(self, nic: dict):
+        """Get the nic string representation."""
+        vendor = nic.get("vendor_name") or nic.get("vendor_id") or "Unknown vendor"
+        product = nic.get("product_name") or nic.get("product_id") or "Unknown product"
+        name = nic.get("name") or "Unknown ifname"
+        return f"{vendor} {product} ({name})"
+
+    def _show_sriov_nics(
+        self,
+        console: Console,
+        sriov_nics: list[dict],
+        pci_whitelist: list[dict],
+        excluded_devices: dict[str, list],
+    ):
+        if not sriov_nics:
+            return
+
+        console.print("Found the following SR-IOV capable devices:")
+
+        for nic in sriov_nics:
+            whitelisted, physnet = self._is_sriov_nic_whitelisted(
+                nic, pci_whitelist, excluded_devices
+            )
+            checkbox = "X" if whitelisted else " "
+            nic_str_repr = self._get_nic_str_repr(nic)
+
+            nic_info = f"  \\[{checkbox}] {nic_str_repr} \\[physnet: {physnet}]"
+            console.print(nic_info)
+
+    def has_prompts(self) -> bool:
+        """Returns true if the step has prompts that it can ask the user.
+
+        :return: True if the step can ask the user for prompts,
+                 False otherwise
+        """
+        return True
+
+    def run(self, status: Status | None = None) -> Result:
+        """Apply individual hypervisor settings."""
+        app = "openstack-hypervisor"
+        action_cmd = "set-hypervisor-local-settings"
+        name = self.node_name
+
+        self.update_status(status, f"setting PCI configuration for {name}")
+
+        excluded_devices = self.variables.get("excluded_devices") or {}
+        node_excluded_devices = excluded_devices.get(name) or []
```

Comment:
If node_excluded_devices is empty list, no need to run action

Reply:
Actually, we do need to apply empty lists. For example, the user may be updating the PCI configuration and enable devices that were previously excluded.